File: primes/src/primes_serial.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

int main(int argc, char *argv[])
{
  unsigned long long int i,j,limit,Nprimes;
  bool *primes;
  time_t start_t,end_t;

  limit=500000000;

  start_t=clock();

  primes = malloc(sizeof(bool)*limit);

  /* fill boolean array */
  for (i=2;i<limit;i++)
    {
      primes[i]=1;
    }
  primes[0]=0;
  primes[1]=0;
  /* Sieve of Eratosthenes */
  for (i=2;i<limit;i++)
    {
      if (primes[i])
	{
	  for (j=i;i*j<limit;j++)
	    {
	      primes[i*j]=0;
	    }
	}
    }

  Nprimes=0;
  for(i=0;i<limit;i++)
    {
      if (primes[i]) Nprimes++;
    }
  
  end_t=clock();
  printf("Nprimes = %llu\n",Nprimes);
  printf("Took %g s\n",(double)(end_t - start_t)/CLOCKS_PER_SEC);
  return EXIT_SUCCESS;
}

File: primes/src/primes_serial.f90
PROGRAM primes_serial
  IMPLICIT NONE
  INTEGER(kind=8) ::  i, Nprimes
  REAL(kind=8) :: start_t, end_t
  LOGICAL, ALLOCATABLE :: primes(:)
  INTEGER, PARAMETER :: limit = 500000000

  CALL cpu_time(start_t)

  ! Initialise 
  ALLOCATE(primes(limit))
  primes=.true.
  primes(1)=.false.
  
  ! Sieve of Eratosthenes
  DO i=2,limit
     IF (primes(i)) THEN
        primes(i*i : limit : i) = .false.
     END IF
  END DO

  ! Count up number of primes
  Nprimes=0
  DO i=1,limit
     IF (primes(i)) Nprimes = Nprimes+1
  END DO
  CALL cpu_time(end_t)

  WRITE(*,*) "Number of primes = ",Nprimes
  WRITE(*,*) "Expended wall clock time = ",end_t-start_t  
  
END PROGRAM primes_serial
